### ISSUE TITLE
Fairness in background worker queues

### ISSUE THREAD

**Issue by eseliger** [sourcegraph/sourcegraph#26895]
We started introducing background workers a while ago, to offload heavy lifting from request paths and to better manage all our background tasks. This is super helpful and has matured well since its introduction. As we start having more workloads there, it becomes imminent that we sometimes run into fairness problems.
- Code intel uploads are sometimes not processing for a while, because the sourcegraph/sourcegraph repo sometimes fills the queue with lots of jobs
- Code intel auto-indexing jobs don't run in a "fair" order, just by creation date. If someone triggers 1000 index jobs for their repos on various branches manually, they will take the executor resources for all users of Sourcegraph for a while
- Batch changes reconciler runs FIFO, if someone starts a batch change with 500 changesets, they all need to process before user B's first changeset even starts processing
- SSBC executions easily add hundreds of jobs into the batches executor queue, if user A starts one of these executions, user B needs to wait for ALL the jobs from user A to finish before their first job even starts
- .. and more examples, which I don't all want to capture here, but you get the idea.

Instead, we should have a more sophisticated dequeueing algorithm, that is more "fair". 
**The ideal end state of this would be: a "fairness group" (TBD) can always have a job start as soon as they queue it, when it's tip of their individual queue.**
In executor terms that means: We always hold so many executor workers warm that the time between "job record becomes tip of individual queue" and "started_at is set (is dequeued)" is close to 0s. We will need to find good values for the fairness triggers as we go and constantly adjust until we come close to the desired end state.

Some considerations for fairness:
- We introduce "fairness groups". These could simply be namespaces, or just users.
- We introduce a maximum concurrency per fairness group
- We do a round-robin of all requesting fairness groups so that ideally every fairness group can always have at least 1 job running
- Later we could also limit the maximum amount of jobs in a given time frame (leaking bucket) so that users can have rather high concurrency but not for an entire hour in a row

This needs to be implemented
- in the dbworker Dequeue method
- in the `rank` virtual field on records, so we can return `placeInQueue` in the API. This should be per individual "fairness group" queue, rather than global. We might want to have a site-admin-only globalPlaceInQueue for debugging purposes.

To experiment with fairness, we want to
- Make maximum concurrency configurable
- Make maximum concurrency overwriteable per sub-queue
- Make this an opt-in feature per workerutil.Worker
- Have a graph/metric to track the dequeue time for tip of queue for each sub queue

In the future, we might allow executors to attach to these "sub queues" for a single fairness group. (Global vs scoped executors). Also, we might want to experiment with "high priority items", ie. user-invoked vs background-job-invoked.

**Comment by eseliger** [comment]
Just hacked around a bit, posting 30 min timebox progress here. Does part of the job so far

```sql
WITH tenant_queues AS (
SELECT
	b.user_id,
	COUNT(1) FILTER (WHERE state = 'queued') as queue_length,
	COUNT(1) FILTER (WHERE state = 'processing') as current_concurrency,
	MAX(j.started_at) as latest_dequeue
FROM batch_spec_workspace_execution_jobs j
JOIN batch_spec_workspaces w ON w.id = j.batch_spec_workspace_id
JOIN batch_specs b ON b.id = w.batch_spec_id
GROUP BY b.user_id
)
SELECT
	q.user_id, q.queue_length, q.current_concurrency, q.latest_dequeue, j.id as job_id
FROM batch_spec_workspace_execution_jobs j
JOIN batch_spec_workspaces w ON w.id = j.batch_spec_workspace_id
JOIN batch_specs b ON b.id = w.batch_spec_id
JOIN tenant_queues q ON q.user_id = b.user_id
WHERE
	-- Limit concurrency.
	q.current_concurrency < 4 AND
	j.state = 'queued' AND
	(j.process_after IS NULL OR j.process_after <= NOW() - interval '2 days')
ORDER BY
	-- Round-robin let tenants dequeue jobs
	ROW_NUMBER() OVER (
		PARTITION BY q.user_id
		ORDER BY
			-- Still dequeue in order of creation (todo: readd order by crated_at, process_after)
			j.id,
			-- TODO: Doesn't work. This should enforce that user 2 can dequeue a job first.
			q.latest_dequeue ASC NULLS FIRST
	);
```

**Comment by rrhyne** [comment]
> Also, we might want to experiment with "high priority items", ie. user-invoked vs background-job-invoked.

I like this, it would allow a user to set priority for critical updates ahead of any other jobs on a busy system. It seems like a feature we might want to wait on building until we hear the need for it though. 

Should we consider allowing a user to set a low priority as well? That would be less troublesome than users being able to set any item to high priority and promote a 'good citizen behavior.

**Comment by unknwon** [comment]
> > Also, we might want to experiment with "high priority items", ie. user-invoked vs background-job-invoked.
> 
> I like this, it would allow a user to set priority for critical updates ahead of any other jobs on a busy system. It seems like a feature we might want to wait on building until we hear the need for it though.

FYI permissions syncing is using this simple {High, Low} priority system for processing requests, where only user-invoked requests can have High priority.

> 
> Should we consider allowing a user to set a low priority as well? That would be less troublesome than users being able to set any item to high priority and promote a 'good citizen behavior.

..., and kinda avoided this problem because only site admins may manually trigger a perms sync 😅

**Comment by eseliger** [comment]
Additional thoughts from code intel:

- We want to introduce "system jobs" that have no specific tenants

**Comment by eseliger** [comment]
We want this to support "high priority" items as well

**Comment by eseliger** [comment]
This will be _deep_ backlog until we revisit our scheduler. A basic version of this has been implemented. The additional constraining features would be most useful for multi-queue executors or multi-tenant executor pools.
